// application/model.js
/* 
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

var mysql      = require('mysql');
var executeQuery = function(query,callback){
  var connection = mysql.createConnection({
    host     : 'localhost',
    user     : 'root',
    password : '',
    database : 'sistemidistribuiti',
    insecureAuth:true
  });

  connection.connect();
  connection.query(query, function(err, rows, fields) {
    if (err) throw err;
    callback(rows);
  });
  connection.end();
  
}

var getParam = function(req,param) {
  console.log(escape(req.query[param]));
  return(escape(req.query[param]));
}
exports.getParam = getParam;

/**
 * @param data : 
 * var data = new Object;
 * data.select = ["username","password"];
 * data.from = "users";
 * data.where = [
 * {
   * 0: "username",
   * 1: "=",
   * 2: "andrea"
 * }
 * ];
 */
var searchQuery =  function(data){
  query = "";
  
  //select
  length = data.select.length;
  query += "SELECT ";
  for(i=0;i<length;i++){
    query += data.select[i]
    if(i<length-1)
      query += ", ";
    else query += " ";
  }
  //from
  query += " FROM "+ data.table;
  //where
  query += " WHERE ";
  length = data.where.length;
  for(i=0;i<length;i++){
    query += data.where[i][0]+data.where[i][1]+"'"+data.where[i][2]+"'";
    if(i<length-1)
      query +=" AND ";
  
  }
  return query;
}

  
var insertQuery = function(data){
  var query = "INSERT INTO "+data.tableName+" (";
  var length = data.fields.length;
  for(var i=0;i<length;i++){
    query += data.fields[i];
    if(i<length-1)
      query += ",";
  }
  query += ") VALUES (";
  
  for(var i=0;i<length;i++){
    if(data.values[i].match(/SELECT/))
      query += "("+data.values[i]+")";
    else
      query += "'"+data.values[i]+"'";
    if(i<length-1)
      query +=",";
  }
  query += ")";
  console.log(query)
  return query;
    
}

/**
   * Crea la query per l'aggiornamento di determinati campi, in una determinata tabella
   * sotto determinate condizioni
   * @param string $table
   * @param array $valueSet i valori da aggiornare array[i]['campo'] array[i]['valore nuovo']
   * @param array $where condizioni che identificano le ennuple da aggiornare
   * @return string la query per l'update
   */
  var updateQuery = function(data){
    var query = "UPDATE " + data.table + " ";
    query += "SET ";
    length = data.valueSet.length;
    for (var i=0; i<length; i++) {
      query += data.valueSet[i][0] + "=" + "'" + data.valueSet[i][1] + "'";
      if (i < length - 1)
        query += ",";
      else
        query += "";
    }
    query += " WHERE ";
    length = data.where.length;
    for (var i=0; i<length; i++) {
      if (!data.where[i][1].match(/IN/))
        query += data.where[i][0] + data.where[i][1] + "'" + data.where[i][2] + "'";
      else
        query += data.where[i][0] + " " + data.where[i][1] + " " + "(" + data.where[i][2] + ")";
      if (i < length - 1)
        query += " AND ";
    }
    return query;
  }

exports.insertQuery = insertQuery;
exports.searchQuery = searchQuery;
exports.executeQuery = executeQuery;
exports.updateQuery = updateQuery;
